bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        its log density; if rejected, propose a new sample and repeat. Stop when the
+        maximum number of proposals is reached or when probabilistic delayed rejection
+        terminates early.
+
+        Flip momemntum after each proposal to ensure detailed balance. Flip momentum
+        before return statement for momentum persistence in Generalized HMC.
+
+        Returns:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        self._rho = self._rng.normal(
+            loc=self._rho * np.sqrt(1 - self._dampening),
+            scale=self._dampening,
+            size=self._dim,
+        )
+        logp_cur = self.joint_logp(self._theta, self._rho)
+        hastings_cur, reject_logp = 0.0, 0.0
+
+        for k in range(self._num_proposals):
+            pdr = self._prob_of_delayed_rejection(reject_logp)
+            if not np.log(self._rng.uniform()) < pdr:
+                break
+
+            stepsize, steps = self._stepsize_list[k], self._steps_list[k]
+            theta_prop, rho_prop = self.leapfrog(
+                self._theta, self._rho, stepsize, steps
+            )
+            rho_prop = -rho_prop
+
+            accept_logp, logp_prop = self.accept(
+                theta_prop, rho_prop, k, hastings_cur, logp_cur
+            )
+            if np.log(self._rng.uniform()) < accept_logp:
+                self._theta, self._rho = theta_prop, rho_prop
+                logp_cur = logp_prop
+                break
+
+            reject_logp = np.log1p(-np.exp(accept_logp))
+            hastings_cur += reject_logp
+        self._rho = -self._rho  # always flip momentum in GHMC
+        return self._theta, logp_cur
+
+    def accept(
+        self,
+        theta_prop: VectorType,
+        rho_prop: VectorType,
+        k: int,
+        hastings_cur: float,
+        logp_cur: float,
+    ) -> tuple[float, float]:
+        """Log acceptance probability of transitioning from current to proposed sample.
+
+        Recursively compute log probability density of accepting transition from
+        current sample (theta, rho) to proposed sample (theta_prop, rho_prop).
+        Information about the current sample is contained in logp_cur and is *not*
+        passed into the function directly via (theta, rho) for efficient reuse.
+
+        Acceptance probability for the k-th proposed sample, with probabilistic delayed
+        rejection, is computed by extending eqns. 29, 30, and 31 of Modi et al. The
+        acceptance probability contains three terms for the current and proposed
+        samples respectively:
+
+            1. the joint log density
+            2. the hastings term for asymmetric proposal kernels; simplfies to the probability of rejecting previous/ghost proposals
+            3. the probability of proposing another sample upon rejection
+
+        Optimize computation by passing hastings_cur and logp from previous proposal
```

Comment:
You don't need all the computational details in the doc.  The doc should be about what a function does, not how it does it.

Reply:
Should I put these computational details in comments in the function itself or just delete them?

I am leaning towards shortening them and putting them as comments in the function itself.

Reply:
Documenting implementation choices as inline comments is often helpful. Personally I don't mind seeing comments for anything overly clever or non-obvious.

Unfortunately, determining what is non-obvious is both subjective and, well, non-obvious. :smiling_face_with_tear:

The balance kind of comes down to a) how thoroughly you expect people who read the code to already know pretty much what you're doing and why; b) how expressive the variable/function names already are; c) how far away the other impacted parts of the code are; and d) whether you can add the inline comments in a way that doesn't totally break up the flow of reading the code itself.

You can reduce the impact of d) a little by writing the comment so it explains the whole "clever" section upfront, rather than doing it interleaved or line-by-line beside the code--with the risk that you're now adding more lines that separate the part you're documenting from the rest of the function. (In the extreme case, this may be an indication that you can separate that part as its own function.)

In short--it's always gonna be a judgment call. We can workshop it with you if you want, or whatever you choose is probably fine.

Reply:
I agree with @jsoules and want to point out the key phrases are "overly clever" and "non-obvious".  If it's obvious to someone skilled in the art of the programming language you're using, don't comment it.  Otherwise, it should be commented.